resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -111,66 +109,104 @@ private[spark] class Client(
+                        .withLabels(resolvedSelectors)
+                        .endMetadata()
+                      .withNewSpec()
+                        .withType("NodePort")
+                        .withSelector(resolvedSelectors)
+                        .withPorts(driverLauncherServicePort)
+                        .endSpec()
+                      .done()
+                    try {
+                      sparkConf.set("spark.kubernetes.driver.service.name",
+                        service.getMetadata.getName)
+                      sparkConf.setIfMissing("spark.driver.port", DEFAULT_DRIVER_PORT.toString)
+                      sparkConf.setIfMissing("spark.blockmanager.port",
+                        DEFAULT_BLOCKMANAGER_PORT.toString)
+                      val driverLauncher = getDriverLauncherClient(kubernetesClient, service)
+                      val ping = Retry.retry(5, 5.seconds) {
+                        driverLauncher.ping()
+                      }
+                      ping onFailure {
+                        case t: Throwable =>
                           submitCompletedFuture.setException(t)
+                          kubernetesClient.services().delete(service)
+                      }
+                      val submitComplete = ping.flatMap { _ =>
+                        Future {
+                          sparkConf.set("spark.driver.host", pod.getStatus.getPodIP)
+                          val submitRequest = buildSubmissionRequest()
+                          driverLauncher.create(submitRequest)
                         }
-                    }
-                    val submitComplete = ping andThen {
-                      case Success(_) =>
-                        driverLauncher.create(submitRequest)
-                        submitCompletedFuture.set(true)
-                    }
-                    submitComplete onFailure {
-                      case t: Throwable =>
-                        if (!submitCompletedFuture.isDone) {
+                      }
+                      submitComplete onFailure {
+                        case t: Throwable =>
                           submitCompletedFuture.setException(t)
+                          kubernetesClient.services().delete(service)
+                      }
+                      val adjustServicePort = submitComplete.flatMap { _ =>
+                        Future {
+                          // After submitting, adjust the service to only expose the Spark UI
+                          val uiServicePort = new ServicePortBuilder()
+                            .withName(UI_PORT_NAME)
+                            .withPort(uiPort)
+                            .withNewTargetPort(uiPort)
+                            .build()
+                          kubernetesClient.services().withName(kubernetesAppId).edit()
+                            .editSpec()
+                              .withType("ClusterIP")
+                              .withPorts(uiServicePort)
```

Comment:
ideally we could just remove the now-unnecessary port rather than resetting back to this state.  Possibly an admin may have added additional ports in the meantime, that this would now override?

Alternatively we can save the "additional app ports" feature for a separate PR since this is already a larger one